app/src/main/java/com/davidtomas/taskyapp/features/auth/presentation/login/LoginViewModel.kt

```diff
@@ -1,5 +1,87 @@
 package com.davidtomas.taskyapp.features.auth.presentation.login
 
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.setValue
 import androidx.lifecycle.ViewModel
+import com.davidtomas.taskyapp.core.domain.util.EMPTY_STRING
+import com.davidtomas.taskyapp.core.domain.util.Failure
+import com.davidtomas.taskyapp.features.auth.domain.model.AllFieldsLoginValidationType
+import com.davidtomas.taskyapp.features.auth.domain.model.EmailFormatError
+import com.davidtomas.taskyapp.features.auth.domain.model.EmailLoginValidationType
+import com.davidtomas.taskyapp.features.auth.domain.model.LoginInputValidationErrors
+import com.davidtomas.taskyapp.features.auth.domain.model.LoginValidationType
+import com.davidtomas.taskyapp.features.auth.domain.model.MissingMandatoryField
+import com.davidtomas.taskyapp.features.auth.domain.model.PasswordFormatError
+import com.davidtomas.taskyapp.features.auth.domain.model.PasswordLoginValidationType
+import com.davidtomas.taskyapp.features.auth.domain.useCase.ValidateLoginInputsUseCase
+
+class LoginViewModel(
+    private val validateLoginInputsUseCase: ValidateLoginInputsUseCase
+) : ViewModel() {
+
+    var state by mutableStateOf(LoginState())
+        private set
+
+    fun onAction(action: LoginAction) {
+        when (action) {
+            is OnChangePasswordVisibility -> {
+                state = state.copy(isPasswordVisible = !state.isPasswordVisible)
+            }
+
+            is OnEmailChanged -> {
+                state = state.copy(email = action.email)
+                validateInputs(EmailLoginValidationType(action.email))
+            }
+
+            is OnPasswordChanged -> {
+                state = state.copy(password = action.password)
+                validateInputs(PasswordLoginValidationType(action.password))
+            }
+
+            else -> Unit
+        }
+    }
+
+    private fun validateInputs(
+        loginValidationType: LoginValidationType
+    ) {
+        validateLoginInputsUseCase.invoke(loginValidationType).fold(
+            fnL = ::onValidationError,
+            fnR = {
```

Comment:
I'd be veeery careful with such abbreviations in the code :D 
I have no clue what `fnL` stands for. I can assume the `L` stands for left, but other than that, it's just cryptic. I highly recommend to only abbreviate what's absolutely obvious, like `ID` or `UI`. For the rest, you're much better off writing longer names. In this case, for example `onError` and `onSuccess`